### ISSUE TITLE
[cs] Default params in haxe dll used from cs

### ISSUE THREAD

**Issue by azrafe7** [HaxeFoundation/haxe#4147]
Having troubles using this class (compiled as a .net 40 dll) from cs:

``` haxe
class IntPoint {

    public var x:Int;
    public var y:Int;

    public function new(x:Int = 0, y:Int = 0) {
        this.x = x;
        this.y = y;
    }
}
```

In CS I get:

``` c#
...
namespace ConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            IntPoint ip0 = new IntPoint();               // 'IntPoint' does not contain a constructor that takes 0 arguments
            IntPoint ip1 = new IntPoint(0, 0);           // cannot convert from 'int' to 'haxe.lang.Null<int>'
            IntPoint ip2 = new IntPoint((int)0, (int)0); // cannot convert from 'int' to 'haxe.lang.Null<int>'
        }
    }
}
```

Am I doing something wrong?

I'm using latest haxe nightly (3.2.0 (git build development @ ec44565)), and git hxcs (also tried with the haxelib ver), compiling with VS Express 2010.

**Comment by nadako** [comment]
Haxe's optional types become nullables from caller's side which is why they are generated as `Null<int>` (or just `object` when `-D erase-generics`), so you'll have to do `new Null<int>(0, true)`.

@waneck Could we maybe add implicit cast function there for better interop (so one could automatically cast from e.g. `int` to `Null<int>`).

**Comment by azrafe7** [comment]
Ok, thanks!

That implicit cast you suggest would be really nice to have, as things can get verbose very fast without it.

Would it be feasible to support _native_ default params when targeting C# 4.0?

**Comment by waneck** [comment]
I've reverted the code that fixed it since it doesn't play well with all optional arguments in Haxe. I'm not sure if it makes sense still to support this only in some cases

**Comment by RealyUniqueName** [comment]
This is fixed with #4623.
Now adding `@:nativeGen` to a class makes the compiler generate overloads for optional arguments. 
The sample in start would work like this:
```haxe
@:nativeGen
class IntPoint {
```
```cs
namespace ConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            IntPoint ip0 = new IntPoint();       
            //or 
            IntPoint ip1 = new IntPoint(new Null<int>(123, true));
            //or 
            IntPoint ip2 = new IntPoint(new Null<int>(123, true), new Null<int>(123, true));
        }
    }
}
```
